lib/saml/settings_service.rb

```diff
@@ -1,48 +1,67 @@
 # frozen_string_literal: true
+require 'memoist'
+
 module SAML
   # This class is responsible for putting together a complete ruby-saml
   # SETTINGS object, meaning, our static SP settings + the IDP settings
   # which must be fetched once and only once via IDP metadata.
   class SettingsService
-    attr_reader :saml_settings
+    class << self
+      extend Memoist
 
-    METADATA_URI = URI(SAML_CONFIG['metadata_url'])
+      METADATA_RETRIES = 3
+      OPEN_TIMEOUT = 2
+      TIMEOUT = 15
 
-    def initialize
-      @saml_settings = create_settings_from_metadata
-    end
+      def saml_settings
+        OneLogin::RubySaml::IdpMetadataParser.new.parse(metadata, settings: settings)
+      rescue => e
+        Rails.logger.error "SAML::SettingService failed to parse SAML metadata: #{e.message}"
+        raise e
+      end
+      memoize :saml_settings
 
-    def self.metadata
-      return @metadata if defined?(@metadata)
-      http = Net::HTTP.new(METADATA_URI.host, METADATA_URI.port)
-      http.use_ssl = true
-      http.verify_mode = OpenSSL::SSL::VERIFY_PEER
-      get = Net::HTTP::Get.new(METADATA_URI.request_uri)
-      @metadata = http.request(get).body
-    end
+      private
 
-    private
+      def connection
+        Faraday.new(SAML_CONFIG['metadata_url']) do |conn|
+          conn.options.open_timeout = OPEN_TIMEOUT
+          conn.options.timeout = TIMEOUT
+          conn.adapter :net_http
+        end
+      end
+      memoize :connection
 
-    def settings
-      # populate with SP settings
-      settings = OneLogin::RubySaml::Settings.new
-      settings.certificate  = SAML_CONFIG['certificate']
-      settings.private_key  = SAML_CONFIG['key']
-      settings.issuer       = SAML_CONFIG['issuer']
-      settings.assertion_consumer_service_url = SAML_CONFIG['callback_url']
+      def metadata
+        attempt ||= 0
+        response = connection.get
+        raise SAML::InternalServerError, response.status if (400..504).cover? response.status.to_i
+        response.body
+      rescue StandardError => e
+        Rails.logger.error "Failed to load SAML metadata: #{e.message}: try #{attempt + 1} of #{METADATA_RETRIES}"
```

Comment:
`Rails.logger.error` could be harsh for attempts 1 & 2.    The whole premise of this PR is that a call or two to get metadata _might_ fail and we recover by trying again.  I wouldn't want an `ERROR` level message to create a red herring in the logs in the event that the first call failed but we successfully recovered in the 2nd call.  What about this instead:

```ruby
msg = "Failed to load SAML metadata: #{e.message}: try #{attempt + 1} of #{METADATA_RETRIES}"
(attempt >= METADATA_RETRIES) ? Rails.logger.error msg : Rails.logger.warn msg
```